File: iosApp/iosApp/Modules/MnemonicConfirmation ✅/MnemonicConfirmationWireframeFactory.swift
// Created by web3d4v on 12/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol MnemonicConfirmationWireframeFactory {
    func make(_ parent: UIViewController?) -> MnemonicConfirmationWireframe
}

final class DefaultMnemonicConfirmationWireframeFactory {
    private let keyStoreService: KeyStoreService
    private let actionsService: ActionsService
    private let networksService: NetworksService
    private let mnemonicService: MnemonicService
    
    init(
        keyStoreService: KeyStoreService,
        actionsService: ActionsService,
        networksService: NetworksService,
        mnemonicService: MnemonicService
    ) {
        self.keyStoreService = keyStoreService
        self.actionsService = actionsService
        self.networksService = networksService
        self.mnemonicService = mnemonicService
    }
}

extension DefaultMnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory {
    
    func make(_ parent: UIViewController?) -> MnemonicConfirmationWireframe {
        DefaultMnemonicConfirmationWireframe(
            parent,
            keyStoreService: keyStoreService,
            actionsService: actionsService,
            networksService: networksService,
            mnemonicService: mnemonicService
        )
    }
}

final class MnemonicConfirmationWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> MnemonicConfirmationWireframeFactory in
            DefaultMnemonicConfirmationWireframeFactory(
                keyStoreService: resolver.resolve(),
                actionsService: resolver.resolve(),
                networksService: resolver.resolve(),
                mnemonicService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicconfirmation/MnemonicConfirmationWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicconfirmation

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.mnemonicConfirmation.MnemonicConfirmationWireframe
import com.sonsofcrypto.web3walletcore.services.actions.ActionsService
import com.sonsofcrypto.web3walletcore.services.mnemonic.MnemonicService
import smartadapter.internal.extension.name

interface MnemonicConfirmationWireframeFactory {
    fun make(parent: Fragment?): MnemonicConfirmationWireframe
}

class DefaultMnemonicConfirmationWireframeFactory(
    private val keyStoreService: KeyStoreService,
    private val actionsService: ActionsService,
    private val mnemonicService: MnemonicService,
): MnemonicConfirmationWireframeFactory {

    override fun make(
        parent: Fragment?
    ): MnemonicConfirmationWireframe = DefaultMnemonicConfirmationWireframe(
        parent?.let { WeakRef(it) },
        keyStoreService,
        actionsService,
        mnemonicService,
    )
}

class MnemonicConfirmationWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(MnemonicConfirmationWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultMnemonicConfirmationWireframeFactory(
                it.resolve(KeyStoreService::class.name),
                it.resolve(ActionsService::class.name),
                it.resolve(MnemonicService::class.name),
            )
        }
    }
}